llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -325,23 +325,45 @@ bool RepoPruning::runOnModule(Module &M) {
   }
 
   // Enable this pass support for alias.
-  for (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;) {
-    Module::alias_iterator CurI = I;
-    ++I;
+  // First, check whether a GlobalAlias references a pruned GO. If yes, record
+  // it. Second, replace all recorded aliases by the declaration. These two
+  // steps can't be merged since the second step could change the base object of
+  // the alias. For example:
+  // Assuming: 1) the aliasee of alias 'A' is alias 'B',
+  //           2) the aliasee of alias 'B' is global object 'C',
+  //           3) the global object 'C' is pruned.
+  // If merging two steps, the order in the alias list is 'B' and then 'A'.
+  // Since the global object 'C' is pruned, the alias 'B' is changed from
+  // definition to declaration. Then, since 'B' is changed to declaration, the
+  // base object of the alias 'A' will be changed from 'C' to 'B'. 'A' will not
+  // be removed because 'B' isn't a global object. However, 'A' must point to a
+  // definition but 'B' is not a definition. This will result in the backend
+  // error.
+  // Step 1. Record aliases.
+  SmallVector<GlobalAlias *, 16> Aliases;
+  for (GlobalAlias &A : M.aliases()) {
```

Comment:
Could this be "**const** GlobalAlias &A"?

Reply:
This can’t be `const GlobalAlias &A` since some aliases need to be put into the `SmallVector<GlobalAlias *, 16> Aliases`, which is used by the step 2.

If we change `SmallVector<GlobalAlias *, 16> Aliases` to `SmallVector<const GlobalAlias *, 16> Aliases`, the step 2 would **not** be built.
```
  for (GlobalAlias *A : Aliases) {
    Type *Ty = A->getValueType();
    // Remove the alias from Module. If not removed, the clone name of the alias
    // will be created in the `createDeclaration` function. This might result in
    // a linker error.
    A->removeFromParent();
    // Replace the alias by a declaration.
    A->replaceAllUsesWith(createDeclaration(M, Ty, A));
    delete A;
  }
```
Both functions `removeFromParent` and `replaceAllUsesWith` requires the alias `A` is **not** `const.`

Reply:
Ah yes, I see now. Thanks for checking.